compiler/grainformat/reformat.re

```diff
@@ -0,0 +1,2753 @@
+    Doc.lbrace,
+    Doc.indent(
+      Doc.concat([
+        Doc.line,
+        Doc.join(
+          Doc.concat([Doc.comma, Doc.line]),
+          List.map(
+            (
+              field: (
+                Grain_parsing__Location.loc(Grain_parsing__Identifier.t),
+                Grain_parsing__Parsetree.expression,
+              ),
+            ) => {
+              let (locidentifier, expr) = field;
+              let ident = locidentifier.txt;
+
+              let printed_ident = print_ident(ident);
+              let printed_expr =
+                print_expression(
+                  ~expr,
+                  ~parentIsArrow=false,
+                  ~endChar=None,
+                  ~original_source,
+                  ~parent_loc,
+                );
+              let punned_expr = check_for_pun(expr);
+
+              let pun =
+                switch (printed_ident) {
+                | Text(i) =>
+                  switch ((punned_expr: Doc.t)) {
+                  | Text(e) => i == e
+                  | _ => false
+                  }
+                | _ => false
+                };
+
+              if (!pun) {
+                Doc.group(
+                  Doc.concat([printed_ident, Doc.text(": "), printed_expr]),
+                );
+              } else {
+                Doc.group(printed_ident);
+              };
+            },
+            fields,
+          ),
+        ),
+        if (List.length(fields) == 1) {
+          Doc.comma; // always append a comma as single argument record look like block {data:val}
+        } else {
+          Doc.ifBreaks(Doc.text(","), Doc.nil);
+        },
+      ]),
+    ),
+    Doc.line,
+    Doc.rbrace,
+  ])
+
+and print_type = (p: Grain_parsing__Parsetree.parsed_type) => {
+  switch (p.ptyp_desc) {
+  | PTyAny => Doc.text("AnyType")
+  | PTyVar(name) => Doc.text(name)
+  | PTyArrow(types, parsed_type) =>
+    Doc.concat([
+      Doc.group(
+        if (List.length(types) == 1) {
+          print_type(List.hd(types));
+        } else {
+          Doc.concat([
+            Doc.lparen,
+            Doc.join(
+              Doc.concat([Doc.comma, Doc.space]),
+              List.map(t => print_type(t), types),
+            ),
+            Doc.rparen,
+          ]);
+        },
+      ),
+      Doc.text(" -> "),
+      print_type(parsed_type),
+    ])
+
+  | PTyTuple(parsed_types) =>
+    add_parens(
+      Doc.join(Doc.comma, List.map(t => print_type(t), parsed_types)),
+    )
+
+  | PTyConstr(locidentifier, parsedtypes) =>
+    let ident = locidentifier.txt;
+    if (List.length(parsedtypes) == 0) {
+      print_ident(ident);
+    } else {
+      Doc.concat([
+        print_ident(ident),
+        Doc.text("<"),
+        Doc.join(
+          Doc.concat([Doc.comma, Doc.space]),
+          List.map(typ => {print_type(typ)}, parsedtypes),
+        ),
+        Doc.text(">"),
+      ]);
+    };
+  | PTyPoly(locationstrings, parsed_type) =>
+    Doc.text("/* Formatter error PTyPoly*/")
```

Comment:
Same deal, we should probably refuse to format the file?

Reply:
Rather than halt, I've replaced the original source instead.  This will let the formatter continue on code it doesn't understand.